OSSMDOC-554: Document new cert management settings for Elasticsearch.

JStickler opened the pull request:
Version(s): 4.7 - 4.11

Issue: [OSSMDOC-554](https://issues.redhat.com/browse/OSSMDOC-554)

Preview here -> http://file.bos.redhat.com/jstickle/OSSMDOC-554/distr_tracing/distr_tracing_install/distr-tracing-deploying-jaeger.html#distr-tracing-manage-es-certificates_deploying-distr-tracing-platform 

This PR documents the new certificate management settings for Elasticsearch and adds the new parameters to the storage configuration reference.

Eng review - frzifus
QE review - jkandasa, iblancasa
Peer review - kelbrown20

frzifus commented on modules/distr-tracing-config-storage.adoc:
```diff
@@ -65,7 +65,13 @@ Memory storage is only appropriate for development, testing, demonstrations, and
+* `spec.storage.elasticsearch.doNotProvision` set to `false`
+* `spec.storage.options.es.server-urls` is not defined, that is, there is no connection to an Elasticsearch instance that was not provisioned by the Red Hat Elasticsearch Operator.
+
+When provisioning Elasticsearch, the {JaegerName} Operator sets the Elasticsearch custom resource `name` to the value of `spec.storage.elasticsearch.name` from the Jaeger custom resource.
```
Not sure if its important, but if no `name` is specified, we use `elasticsearch` as default value.

https://github.com/jaegertracing/jaeger-operator/blob/0077e5232ce194843c012cb8605b3470a9eeab7b/apis/v1/jaeger_webhook.go#L52

frzifus commented on modules/distr-tracing-config-storage.adoc:
```diff
@@ -133,6 +153,12 @@ Minimum deployment = 16Gi*
 
+|elasticsearch:
+  useCertManagement:
+|Use to specify whether or not {JaegerShortName} should use the certificate management feature of the Red Hat Elasticsearch Operator.
```
In the comments are a few more details.
```
// Whether Elasticsearch cert management feature should be used.
// This is a preferred setting for new Jaeger deployments on OCP versions newer than 4.6.
// The cert management feature was added to Red Hat Openshift logging 5.2 in OCP 4.7.
```

frzifus commented on modules/distr-tracing-config-storage.adoc:
```diff
@@ -189,7 +215,15 @@ spec:
+
+* `spec.storage.elasticsearch.doNotProvision` set to `true`
+* `spec.storage.options.es.server-urls` has a value
+* `spec.storage.elasticsearch.name` has a value
```
It works without a value if the OpenShift Elasticsearch instance name is `elasticsearch`.

frzifus commented on modules/distr-tracing-config-storage.adoc:
```diff
@@ -615,3 +649,80 @@ spec:
+Starting with version 2.4, the {JaegerName} Operator delegates certificate creation to the Red Hat Elasticsearch Operator by using the following annotations in the Elasticsearch custom resource:
+
+* `logging.openshift.io/elasticsearch-cert-management: true`
+* `logging.openshift.io/elasticsearch-cert.jaeger-shared-es: user.jaeger`
```
We need to add one thing more.
```suggestion
* `logging.openshift.io/elasticsearch-cert.jaeger-shared-es: user.jaeger`
* `logging.openshift.io/elasticsearch-cert.curator-shared-es: system.logging.curator`
```

frzifus commented on modules/distr-tracing-config-storage.adoc:
```diff
@@ -615,3 +649,80 @@ spec:
+    type: elasticsearch
+    elasticsearch:
+      name: elasticsearch
+      doNotProvision: true
```
```
// Whether Elasticsearch cert management feature should be used.
// This is a preferred setting for new Jaeger deployments on OCP versions newer than 4.6.
// The cert management feature was added to Red Hat Openshift logging 5.2 in OCP 4.7.
```

```suggestion
      doNotProvision: true
      useCertManagement: true
```

frzifus reviewed (commented):
Hi @JStickler, Ive left a few comments to the best of my knowledge. :)

iblancasa commented on modules/distr-tracing-config-storage.adoc:
```diff
@@ -65,7 +65,13 @@ Memory storage is only appropriate for development, testing, demonstrations, and
+
+* `spec.storage:type` is set to `elasticsearch`
+* `spec.storage.elasticsearch.doNotProvision` set to `false`
+* `spec.storage.options.es.server-urls` is not defined, that is, there is no connection to an Elasticsearch instance that was not provisioned by the Red Hat Elasticsearch Operator.
```
Would it make sense to just say "is not defined" since you can deploy an ES instance with the ES OpenShift Operator and provide it here? /cc @pavolloffay

JStickler commented:
Added Tech Preview note, as requested by the team (Slack conversation "We would like to make this feature as tech preview for this release (2.4). The reason being is that we don't have a good test coverage for it right now.").  Updated PR based on comments from Benne and updated the preview output.

JStickler commented:
@jkandasa or @iblancasa  I think this one is ready for QE review please.

iblancasa commented:
LGTM. @jkandasa could you review it too? Thanks!

frzifus reviewed (approved):
LGTM :)

jkandasa commented:
@JStickler  I haven't tested the new fields in es section. I will check those and provide my review

jkandasa commented on modules/distr-tracing-config-storage.adoc:
```diff
@@ -133,6 +153,12 @@ Minimum deployment = 16Gi*
+  useCertManagement:
+|Use to specify whether or not {JaegerShortName} should use the certificate management feature of the Red Hat Elasticsearch Operator.  This feature was added to {logging-title} 5.2 in {product-title} 4.7 and is the preferred setting for new Jaeger deployments.
+|`true`/`false`
+|`true`
```
@JStickler @pavolloffay I am bit confused here.
I see you have marked cert-management feature as `tech-preview`, but why do you apply tech-preview feature as default option?
and why do you upgrade the existing installation to new cert management [1]

[1] - https://github.com/jaegertracing/jaeger-operator/blob/v1.34.1/pkg/upgrade/v1_31_0.go

![image](https://user-images.githubusercontent.com/1004403/173055624-246028c2-4a69-487d-a90d-65b108a04bf9.png)

JStickler replied:
@jkandasa see if how I explained it in the [Release Notes PR](https://github.com/openshift/openshift-docs/pull/45350) makes sense.

jkandasa replied:
@JStickler thanks for the Release Notes link. There is a typo there. I have updated in the merged PR :)

frzifus commented on modules/distr-tracing-config-storage.adoc:
```diff
@@ -615,3 +649,80 @@ spec:
+metadata:
+  annotations:
+    logging.openshift.io/elasticsearch-cert-management: true
+    logging.openshift.io/elasticsearch-cert.jaeger-shared-es: user.jaeger
```
@JStickler isn it still missing?

frzifus reviewed (changes requested):
Seems we still miss the curator annotation in the `Example Elasticsearch CR showing annotations`
https://github.com/openshift/openshift-docs/pull/46088/files#r895627971

jkandasa commented on modules/distr-tracing-config-storage.adoc:
```diff
@@ -615,3 +649,90 @@ spec:
+        limits:
+          memory: 16Gi
+      useCertManagment: true
+----
```
@JStickler remove `nodeCount` and `resources` is not taken care by `jaeger`, as `doNotProvision: true`. also remove the `namespace`.

JStickler replied:
@jkandasa Please check and make sure I correctly interpreted which lines to edit.

frzifus replied:
good catch @jkandasa! As far as i can see, it looks fine :+1:

jkandasa replied:
@JStickler typo in 
```
      useCertManagment: true
```
Other than this all other changes looks good to me.

jkandasa reviewed (changes requested):
@JStickler today I get a time to verify es cert manage tech-preview feature, I have added review comments. in addition to that my observations,
* both es node and jaeger instances should be on the same namespace

#### CR used for the tests
Elasticsearch CR. 
**Note: I have restricted resource limit**
```yaml
apiVersion: logging.openshift.io/v1
kind: Elasticsearch
metadata:
  annotations:
    logging.openshift.io/elasticsearch-cert-management: "true"
    logging.openshift.io/elasticsearch-cert.jaeger-custom-es: "user.jaeger"
    logging.openshift.io/elasticsearch-cert.curator-custom-es: "system.logging.curator"
  name: custom-es
spec:
  managementState: Managed
  nodeSpec:
    resources:
      limits:
        memory: 1Gi
        cpu: 1
      requests:
        cpu: 1
        memory: 1Gi
  nodes:
    - nodeCount: 1
      proxyResources: {}
      resources: {}
      roles:
        - master
        - client
        - data
      storage: {}
  redundancyPolicy: ZeroRedundancy
```
Jaeger CR:
```yaml
apiVersion: jaegertracing.io/v1
kind: Jaeger
metadata:
  name: jaegerprod
spec:
  strategy: production
  storage:
    type: elasticsearch
    elasticsearch:
      name: custom-es
      doNotProvision: true
      useCertManagement: true
```

#### OCP cluster version and operators used
```
$ oc version
Client Version: 4.7.0-202106252127.p0.git.8b4b094-8b4b094
Server Version: 4.9.0-0.nightly-2022-06-14-225707
Kubernetes Version: v1.22.8+f34b40c

$ oc get csv
NAME                           DISPLAY                                          VERSION    REPLACES                  PHASE
elasticsearch-operator.5.4.2   OpenShift Elasticsearch Operator                 5.4.2                                Succeeded
jaeger-operator.v1.34.1-5      Red Hat OpenShift distributed tracing platform   1.34.1-5   jaeger-operator.v1.30.2   Succeeded
```

jkandasa reviewed (approved):
Thanks @JStickler! LGTM :)

kelbrown20 commented on modules/distr-tracing-config-storage.adoc:
```diff
@@ -133,6 +153,12 @@ Minimum deployment = 16Gi*
 
+|elasticsearch:
+  useCertManagement:
+|Use to specify whether or not {JaegerShortName} should use the certificate management feature of the Red Hat Elasticsearch Operator.  This feature was added to {logging-title} 5.2 in {product-title} 4.7 and is the preferred setting for new Jaeger deployments.
```
Same thing for the `whether or not` here

kelbrown20 commented on modules/distr-tracing-config-storage.adoc:
```diff
@@ -615,3 +649,84 @@ spec:
+[id="distr-tracing-manage-es-certificates_{context}"]
+= Managing certificates with Elasticsearch
+
+You can create and manage certificates using the Red Hat Elasticsearch Operator. Managing certificates using the Red Hat Elasticsearch Operator also lets you use a single Elasticsearch cluster with multiple Jaeger Collectors.
```
Also up to you!

kelbrown20 commented on modules/distr-tracing-config-storage.adoc:
```diff
@@ -615,3 +649,84 @@ spec:
+* {product-title} 4.7
+* {logging-title} 5.2
+* The Elasticsearch node and the Jaeger instances must be deployed in the same namespace.  For example, `tracing-system`.
+
```
This might just be a me thing that I'm confused about. Are these the prerequisites to the "setting `spec.storage.elasticsearch.useCertManagement` to true" with the example?

JStickler replied:
These are the prerequisites to being able to use the Certificate Manager feature (the feature isn't available before OCP 4.7 or Logging 5.2, and the Jaeger and Elasticsearch components have to be in the same namespace).

kelbrown20 replied:
Oh I see it now! I think I was just confusing myself as I was reading. Thank you!

kelbrown20 commented on modules/distr-tracing-config-storage.adoc:
```diff
@@ -615,3 +649,84 @@ spec:
+
+You can create and manage certificates using the Red Hat Elasticsearch Operator. Managing certificates using the Red Hat Elasticsearch Operator also lets you use a single Elasticsearch cluster with multiple Jaeger Collectors.
+
+[IMPORTANT]
```
I believe this TP preview block is meant to use the TP snippet:

https://github.com/openshift/openshift-docs/blob/main/snippets/technology-preview.adoc

JStickler replied:
I know the snippet exists, I just don't use it.

kelbrown20 reviewed (commented):
Hello! Just a few minor nits and a few questions, other that that this looks great! Please let me know if you have any questions on anything I added 🦖

JStickler commented:
/cherry-pick enterprise-4.7

JStickler commented:
/cherry-pick enterprise-4.8

JStickler commented:
/cherry-pick enterprise-4.9

JStickler commented:
/cherry-pick enterprise-4.10

JStickler commented:
/cherry-pick enterprise-4.11
